internal/services/synapse/synapse_pipeline_resource.go

```diff
@@ -0,0 +1,361 @@
+		if err := d.Set("variables", variables); err != nil {
+			return fmt.Errorf("setting `variables`: %+v", err)
+		}
+
+		if activities := props.Activities; activities != nil {
+			activitiesJson, err := serializeSynapsePipelineActivities(activities)
+			if err != nil {
+				return fmt.Errorf("serializing `activities_json`: %+v", err)
+			}
+			if err := d.Set("activities_json", activitiesJson); err != nil {
+				return fmt.Errorf("setting `activities_json`: %+v", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func resourceSynapsePipelineDelete(d *schema.ResourceData, meta interface{}) error {
+	synapseClient := meta.(*clients.Client).Synapse
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+	environment := meta.(*clients.Client).Account.Environment
+
+	id, err := parse.PipelineID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	client, err := synapseClient.PipelinesClient(id.WorkspaceName, environment.SynapseEndpointSuffix)
+	if err != nil {
+		return err
+	}
+
+	if _, err = client.DeletePipeline(ctx, id.Name); err != nil {
+		return fmt.Errorf("deleting %s: %+v", id, err)
+	}
+
+	return nil
+}
+
+func expandSynapseParameters(input map[string]interface{}) map[string]*artifacts.ParameterSpecification {
+	output := make(map[string]*artifacts.ParameterSpecification)
+
+	for k, v := range input {
+		output[k] = &artifacts.ParameterSpecification{
+			Type:         artifacts.ParameterTypeString,
+			DefaultValue: v.(string),
+		}
+	}
+
+	return output
+}
+
+func flattenSynapseParameters(input map[string]*artifacts.ParameterSpecification) map[string]interface{} {
+	output := make(map[string]interface{})
+
+	for k, v := range input {
+		if v != nil {
+			// we only support string parameters at this time
+			val, ok := v.DefaultValue.(string)
+			if !ok {
+				log.Printf("[DEBUG] Skipping parameter %q since it's not a string", k)
+			}
+
+			output[k] = val
+		}
+	}
+
+	return output
+}
+
+func flattenSynapseAnnotations(input *[]interface{}) []string {
+	annotations := make([]string, 0)
+	if input == nil {
+		return annotations
+	}
+
+	for _, annotation := range *input {
+		val, ok := annotation.(string)
+		if !ok {
+			log.Printf("[DEBUG] Skipping annotation %q since it's not a string", val)
+		}
+		annotations = append(annotations, val)
+	}
+	return annotations
+}
+
+func expandSynapseVariables(input map[string]interface{}) map[string]*artifacts.VariableSpecification {
+	output := make(map[string]*artifacts.VariableSpecification)
+
+	for k, v := range input {
+		output[k] = &artifacts.VariableSpecification{
+			Type:         artifacts.VariableTypeString,
+			DefaultValue: v.(string),
+		}
+	}
+
+	return output
+}
+
+func flattenSynapseVariables(input map[string]*artifacts.VariableSpecification) map[string]interface{} {
+	output := make(map[string]interface{})
+
+	for k, v := range input {
+		if v != nil {
+			// we only support string parameters at this time
+			val, ok := v.DefaultValue.(string)
```

Comment:
Shall we also nil check the `DefaultValue` here?

Reply:
This is to keep consistent with ADF code [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/b0c897055329438be6a3a159f6ffac4e1ce958f2/internal/services/datafactory/data_factory.go#L70-L82)

Reply:
Talked to @magodo , the code logic is okay because we checked `ok` here